lib/mkFakeDerivation.nix

```
@@ -47,7 +48,7 @@
             cacheUrl: cacheMetadata:
               if cacheUrl != null
               then cacheUrl
-              else if builtins.any (narinfo: narinfo.path == stringOutPath) (cacheMetadata.narinfo or [])
+              else if builtins.any (narinfo: narinfo.path == stringOutPath && narinfo.valid == true) (cacheMetadata.narinfo or [])
```

How much backwards compat do we offer here, i.e. this is a breaking change when working with catalogs which dont have the `valid=false` part.


```suggestion
              else if builtins.any (narinfo: narinfo.path == stringOutPath && (narinfo.valid or false) == true) (cacheMetadata.narinfo or [])
```

---

I can't remember - do any of our catalogs only have valid when it's false? I think because we regenerate catalogs this might be okay. But if not, we could change this to `(narinfo ? valid && narinfo.valid == true) || !(narinfo ? valid)`

---

i think they were missing either `valid = true` or `valid = false`.
nixpkgs catalogs can likely be regenerated, I'm rather speaking of older `published` ones

---

Looks like we've had `valid = true` for a while https://github.com/flox/floxpkgs/blob/master/catalog/x86_64-linux/stable/flox-prerelease/0.1.0-r269-0.0.9-r134.json

---

@ysndr I can't find a catalog for which we need to include the legacy logic - do you think it's worth adding?

---

I think the only place we experience that would be environment's `catalog.json`s for the public catalog we have everything cached.

We did change the cache format once to be closer to nix right?

<details>
<summary>from</summary>

```json
cache: [  
  {
      "cacheUrl": "https://cache.floxdev.com",
      "narinfo": [
        {
          ...
          "valid": true
        }
      ]
  }
]
```

</details>


<details>
<summary>to</summary>

```json
"cache": {
   "bin": {
     "https://cache.nixos.org": {}
   },
   "debug": {
     "https://cache.nixos.org": {}
   }
},
```

</details>


That would make the format without `valid = false` pretty early.

---

I think you got from and to backwards, but yeah. This is only touching the codepath for the newer format (see line 66 for handling the older format). So do you think it's okay to leave as is or want me to make the change?

---

I think we can leave it as is, since this is the newer format

https://github.com/NixOS/nix/pull/7924/files added `valid = true` so valid = false should always be there (i think i got that the wrong way around) and iirc we ensured valid =true is in there when we switched to the new format